tools/pubsys/src/aws/client.rs

```diff
@@ -1,159 +1,151 @@
-        D: DispatchSignedRequest + Send + Sync + 'static,
-    {
-        Self::new_with(request_dispatcher, credentials_provider, region)
-    }
-}
+pub(crate) async fn build_client_config<S1, S2>(
+    region: S1,
+    sts_region: S2,
+    aws: &AwsConfig,
+) -> Result<SdkConfig>
+where
+    S1: AsRef<str>,
+    S2: AsRef<str>,
+{
+    let mut config = aws_config::from_env();
+    let region = region.as_ref().to_owned();
+    let sts_region = sts_region.as_ref().to_owned();
 
-impl NewWith for StsClient {
-    fn new_with<P, D>(request_dispatcher: D, credentials_provider: P, region: Region) -> Self
-    where
-        P: ProvideAwsCredentials + Send + Sync + 'static,
-        D: DispatchSignedRequest + Send + Sync + 'static,
-    {
-        Self::new_with(request_dispatcher, credentials_provider, region)
-    }
-}
+    // Add custom endpoint if specified in aws.region.REGION.endpoint.
+    let maybe_regional_endpoint = aws.region.get(&region).and_then(|r| r.endpoint.clone());
+    config = if let Some(endpoint) = &maybe_regional_endpoint {
+        config.endpoint_resolver(Endpoint::immutable(endpoint.parse().expect("valid URI")))
+    } else {
+        config
+    };
 
-/// Create a rusoto client of the given type using the given region and configuration.
-pub(crate) fn build_client<T: NewWith>(
-    region: &Region,
-    sts_region: &Region,
-    aws: &AwsConfig,
-) -> Result<T> {
-    let maybe_regional_role = aws.region.get(region.name()).and_then(|r| r.role.clone());
-    let assume_roles = aws.role.iter().chain(maybe_regional_role.iter()).cloned();
-    let provider = build_provider(
-        sts_region,
-        assume_roles.clone(),
-        base_provider(&aws.profile)?,
-    )?;
-    Ok(T::new_with(
-        rusoto_core::HttpClient::new().context(error::HttpClientSnafu)?,
-        provider,
-        region.clone(),
-    ))
-}
+    // If the user specified a profile, use that, otherwise use the default
+    // credentials mechanisms.
+    config = if let Some(profile) = &aws.profile {
+        config.credentials_provider(
+            ProfileFileCredentialsProvider::builder()
+                .profile_name(profile)
+                .build(),
+        )
+    } else {
+        config
+    };
 
-/// Wrapper for trait object that implements ProvideAwsCredentials to simplify return values.
-/// Might be able to remove if rusoto implements ProvideAwsCredentials for
-/// Box<ProvideAwsCredentials>.
-struct CredentialsProvider(Box<dyn ProvideAwsCredentials + Send + Sync + 'static>);
-#[async_trait]
-impl ProvideAwsCredentials for CredentialsProvider {
-    async fn credentials(&self) -> std::result::Result<AwsCredentials, CredentialsError> {
-        self.0.credentials().await
-    }
+    // Load a chained credential config if role is specified in aws.region.REGION.role.
+    let maybe_regional_role = aws.region.get(&region).and_then(|r| r.role.clone());
+    let config = if maybe_regional_role.is_some() {
```

Comment:
I'm confused about this `let config`... are we replacing the `config` we just built?

Reply:
Correct, but I can rename it to make it more clear.